File: cortex/autopilot/detectors/ambiguous.py
"""cortex.autopilot.detectors.ambiguous — Ambiguous-request detector."""
from __future__ import annotations

from cortex.autopilot.models import DetectionRequest, DetectionResult


class AmbiguousRequestDetector:
    """Detects vague user requests that need clarification before preflight.

    Heuristics (from §7.1.1 of the global plan):
    - request with fewer than 8 meaningful words
    - vague verbs without concrete object ("mejorar", "fix", etc.)
    - no mention of files, modules or specific functions
    - no implicit acceptance criterion
    """

    name = "ambiguous_request"

    VAGUE_VERBS = {
        "mejorar", "arreglar", "cambiar", "actualizar", "fixear",
        "improve", "fix", "change", "update", "refactor",
    }
    MIN_WORDS = 8
    FILE_EXTS = ("py", "ts", "js", "md", "yaml", "json")

    def detect(self, request: DetectionRequest) -> DetectionResult:
        if not request.user_request:
            return DetectionResult(
                task_type="ambiguous",
                confidence=0.9,
                reason="No user request provided",
                suggested_complexity="none",
            )

        words = request.user_request.lower().split()
        has_vague_verb = any(w in self.VAGUE_VERBS for w in words)
        is_short = len(words) < self.MIN_WORDS
        has_file_ref = any(
            "." in w and w.split(".")[-1] in self.FILE_EXTS
            for w in words
        )

        if is_short and has_vague_verb and not has_file_ref:
            return DetectionResult(
                task_type="ambiguous",
                confidence=0.7,
                reason=f"Short request ({len(words)} words) with vague verb, no file references",
                suggested_complexity="none",
            )

        return DetectionResult(
            task_type="noop",
            confidence=0.0,
            reason="Request appears sufficiently specific",
            suggested_complexity="none",
        )

File: rust/crates/cortex-autopilot/src/detectors/ambiguous.rs
//! `AmbiguousRequestDetector`.

use super::AutopilotDetector;
use crate::models::{DetectionRequest, DetectionResult};

pub struct AmbiguousRequestDetector;
impl AutopilotDetector for AmbiguousRequestDetector {
    fn name(&self) -> &'static str {
        "ambiguous_request"
    }
    fn detect(
        &self,
        request: &DetectionRequest,
    ) -> Result<DetectionResult, cortex_enterprise::error::EnterpriseError> {
        const VAGUE_VERBS: &[&str] = &[
            "mejorar",
            "arreglar",
            "cambiar",
            "actualizar",
            "fixear",
            "improve",
            "fix",
            "change",
            "update",
            "refactor",
        ];
        const MIN_WORDS: usize = 8;
        const FILE_EXTS: &[&str] = &["py", "ts", "js", "md", "yaml", "json"];

        let Some(req) = &request.user_request else {
            return Ok(DetectionResult {
                task_type: "ambiguous".into(),
                confidence: 0.9,
                reason: "No user request provided".into(),
                suggested_complexity: "none".into(),
            });
        };
        let words: Vec<String> = req
            .to_lowercase()
            .split_whitespace()
            .map(String::from)
            .collect();
        let has_vague_verb = words.iter().any(|w| VAGUE_VERBS.contains(&w.as_str()));
        let is_short = words.len() < MIN_WORDS;
        let has_file_ref = words
            .iter()
            .any(|w| w.contains('.') && FILE_EXTS.contains(&w.rsplit('.').next().unwrap_or("")));

        Ok(if is_short && has_vague_verb && !has_file_ref {
            DetectionResult {
                task_type: "ambiguous".into(),
                confidence: 0.7,
                reason: format!(
                    "Short request ({} words) with vague verb, no file references",
                    words.len()
                ),
                suggested_complexity: "none".into(),
            }
        } else {
            DetectionResult::noop("Request appears sufficiently specific")
        })
    }
}
